### PR TITLE
feat: last event resolver

### PR DESCRIPTION
**Pull request by Jarsen136**
WFM

image

ref:
#29 
https://github.com/kodadot/nft-gallery/issues/3059

### CODE REVIEW

**Comment by Jarsen136** [comment]
This PR seems conflict with https://github.com/kodadot/rubick/pull/76 : )
cc @vikiival

**Comment by Jarsen136** [comment]
This PR seems conflict with https://github.com/kodadot/rubick/pull/76 : )
cc @vikiival

**Review by roiLeo — COMMENTED** [review]
would be cool to have interaction from `lastListEventQuery` as a param so we can do the same for 'BUY'

It also lacks the possibility of sorting

File: src/server-extension/resolvers/event.ts [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+  @Query(() => [LastEventEntity])
+  async lastListEvent(
+    @Arg('limit', { nullable: true }) limit: number,
+    @Arg('offset', { nullable: true }) offset: number,
+  ): Promise<LastEventEntity[]> {
+    const result: LastEventEntity[] = await this.genericEventQuery(lastListEventQuery(limit, offset))
+    return result
+  }
+
+  async genericEventQuery(query: string): Promise<LastEventEntity[]> {
```
Comment (vikiival) [inline src/server-extension/resolvers/event.ts:21]: use makeQuery pls

File: src/server-extension/query/event.ts [hunk 2]
```diff
@@ -17,3 +17,23 @@ and ce.id in (${idList})
...
+    me.image as metadata_image,
+    me.id as metadata_id
+FROM event e
+    JOIN nft_entity ne on e.nft_id = ne.id
+    LEFT join metadata_entity me on me.id = ne.metadata
+where
+    e.interaction = 'LIST'
+    and ne.burned = false
+ORDER BY e.timestamp desc
+LIMIT ${limit} OFFSET ${offset}`
```
Comment (vikiival) [inline src/server-extension/query/event.ts:39]:
```suggestion
LIMIT $1 OFFSET $2`
```

File: src/server-extension/resolvers/event.ts [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+
+
+@Resolver()
+export class EventResolver {
+  constructor(private tx: () => Promise<EntityManager>) {}
+
+  @Query(() => [LastEventEntity])
+  async lastListEvent(
+    @Arg('limit', { nullable: true }) limit: number,
+    @Arg('offset', { nullable: true }) offset: number,
```
Comment (vikiival) [inline src/server-extension/resolvers/event.ts:15]: add default to 0

File: src/server-extension/resolvers/event.ts [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+import { lastListEventQuery } from '../query/event'
+
+
+@Resolver()
+export class EventResolver {
+  constructor(private tx: () => Promise<EntityManager>) {}
+
+  @Query(() => [LastEventEntity])
+  async lastListEvent(
+    @Arg('limit', { nullable: true }) limit: number,
```
Comment (vikiival) [inline src/server-extension/resolvers/event.ts:14]: add default to 20

**Review by vikiival — CHANGES_REQUESTED** [review]

File: src/server-extension/model/event.model.ts [hunk 5]
```diff
@@ -26,4 +26,38 @@ export class HistoryEntity {
...
+  issuer!: String
+
+  @Field(() => String, { nullable: false })
+  timestamp!: String
+
+  @Field(() => String, { nullable: false })
+  metadata!: String
+
+  @Field(() => String, { nullable: false })
+  meta!: String
```
Comment (vikiival) [inline src/server-extension/model/event.model.ts:49]: should be named `value`

**Comment by Jarsen136** [comment]
> It also lacks the possibility of sorting

Maybe it should sort by timestamp by default because it was named `lastEvent`

File: src/server-extension/model/event.model.ts (hunk 5 above)
Comment (Jarsen136) [reply src/server-extension/model/event.model.ts]: : )

File: src/server-extension/query/event.ts [hunk 6]
```diff
@@ -17,3 +17,22 @@ and ce.id in (${idList})
...
+    (e.meta::bigint) as value,
+    e.timestamp,
+    e.current_owner,
+    me.image as image
+FROM event e
+    JOIN nft_entity ne on e.nft_id = ne.id
+    LEFT join metadata_entity me on me.id = ne.metadata
+where
+    e.interaction = $1
+    and ne.burned = false
```
Comment (vikiival) [inline src/server-extension/query/event.ts:36]:
```suggestion
    AND ne.burned = false
```

**Comment by Jarsen136** [comment]
How about merging this pull request and deploying it ? 👀

**Review by vikiival — APPROVED** [review]

**Comment by vikiival** [comment]
> How about merging this pull request and deploying it ? 👀

Sorry, we've been busy with the dotsama prague

**Pull request merged**